### ISSUE TITLE
False positive for terraform_deprecated_interpolation

### ISSUE THREAD

**Issue by mgrecar** [terraform-linters/tflint#878]
## Version
TFLint 0.19.0
Terraform 0.12.29

## Issue
I'm getting a false positive with rule `terraform_deprecated_interpolation` using TFLint 0.19.0:
```
❯ tfl
1 issue(s) found:

Warning: Interpolation-only expressions are deprecated in Terraform v0.12.14 (terraform_deprecated_interpolation)

  on main.tf line 6:
   6:   grafana_admin_password   = var.grafana_admin_password

Callers:
   main.tf:6,30-56
   ../../../modules/k8s_cluster/main.tf:39,36-81

Reference: https://github.com/terraform-linters/tflint/blob/v0.19.0/docs/rules/terraform_deprecated_interpolation.md
```

It's incorrectly identifying that variable syntax as the old v0.11 style syntax.  This was working just fine for me with 0.18.0, looks like a regression.

**Comment by mgrecar** [comment]
For clarity, `tfl` is just a shell function that runs a Docker image with TFLint in it, and supplies the `.tflint.hcl` file to the call, it's just a convenience that wraps up all the boilerplate for us.

**Comment by bendrucker** [comment]
Hmm, that's quite odd. Any additional details you could provide from your code would be helpful, since there's at least one test covering a resource:

https://github.com/terraform-linters/tflint/blob/master/rules/terraformrules/terraform_deprecated_interpolation_test.go

**Comment by mgrecar** [comment]
Hmm, okay, I dug a bit more into the situation, and it appears it might be a valid error, but not quite reported in a clear way.  It comes down to the configuration of `module = true` being set.  Below is some code I was able to use to make a minimal reproduction of the issue:

.tflint.hcl
```
config { 
  module = true
  deep_check = true
  force = false
} 
```

main.tf
```
module "test" {
  source = "./module"

  value = var.root_value
}

variable "root_value" {
  type        = string
  description = "a string variable provided at the root module to a submodule"
  default     = "hello!"
} 
```

module/main.tf
```
variable "value" {
  type        = string
  description = "provided by root module"
} 

locals {
  local_variable = "${base64encode(var.value)}"
}                

output "print" {
  description = "an output to use the local for testing"
  value       = local.local_variable
} 
```

TFLint gives me:
```
❯ tfl
1 issue(s) found:

Warning: Interpolation-only expressions are deprecated in Terraform v0.12.14 (terraform_deprecated_interpolation)

  on main.tf line 4:
   4:   value = var.root_value

Callers:
   main.tf:4,11-25
   module/main.tf:7,20-48

Reference: https://github.com/terraform-linters/tflint/blob/v0.19.0/docs/rules/terraform_deprecated_interpolation.md
```
While the immediate text prints out where the module receives the input variable `root_value`, in a Terraform v0.12 compliant format, the `Callers` information points you to `module/main.tf:7`:
```
local_variable = "${base64encode(var.value)}"
```
Which is using the old style interpolation, and I believe can be rewritten as:
```
local_variable = base64encode(var.value)
```
Making that change satisfies TFLint.  

So, while it's found a valid change to be made in the Terraform code, it isn't calling out the actual line in the module with the problem, but instead the location where the problematic variable is set for the module in the root configuration.  I think the issue here is I'd expect to see the following output in this case:
```
❯ tfl
1 issue(s) found:

Warning: Interpolation-only expressions are deprecated in Terraform v0.12.14 (terraform_deprecated_interpolation)

  on module/main.tf line 7:
   7:   local_variable = "${base64encode(var.value)}"

Callers:
   main.tf:4,11-25
   module/main.tf:7,20-48

Reference: https://github.com/terraform-linters/tflint/blob/v0.19.0/docs/rules/terraform_deprecated_interpolation.md
```
That's actually pointing to the code that has the problem.  So, the rule's actually successfully finding the issue, but not reporting the location it's actually occuring at properly when module checking is set to `true`.  

I hope that helps highlight the issue!

**Comment by bendrucker** [comment]
Awesome, that adds up! I'll have to play with this a bit and try to add some more test cases. Appreciate the repro.

**Comment by bendrucker** [comment]
Stepping through this example in a debugger, I've confirmed that the emitted range from the rule is correct.

<img width="348" alt="Screen Shot 2020-08-18 at 17 26 05" src="https://user-images.githubusercontent.com/808808/90578358-f8d09b00-e177-11ea-8a0d-b2c6eac37632.png">

So it would seem that this is not necessarily specific to this rule. Stepping into `EmitIssue`, I noticed:

```go
for _, modVar := range r.listModuleVars(r.currentExpr) {
	r.emitIssue(&Issue{
		Rule:    rule,
		Message: message,
		Range:   modVar.DeclRange,
		Callers: append(modVar.callers(), location),
	})
}
```

That would seem to suggest that this isn't a regression related to Terraform 0.13 or TFLint 0.19, since this behavior has been present for a while, added in https://github.com/terraform-linters/tflint/pull/396. The practical reasons mentioned there are tricky but I agree that the present behavior is not the correct default.

In rules that assert things like syntax (as in deprecated interpolations), you'd want errors reported at the range in the child module. In high level rules like the AWS generated ones, bubbling issues to parent variables seems reasonable, but that would be something each rule/issue should specify on an opt-in basis.

It also seems like the snippet above results in missed issues when using `--module`. When using `--module`, only issues emitted on expressions containing variables will be bubbled up.

A simple repro is:

**module/main.tf**

```tf
locals {
  foo = "bar"
  baz = "${local.foo}"
}
```

**main.tf**

```tf
module "m" {
  source = "./module"
}
```

TFLint exits successfully, since the emitted issue on the expression `"${local.foo}"` references no variables and therefore is never emitted. 

Currently `--module` is effectively trying to provide validation of the module _arguments_ (attributes). But that doesn't make sense for generic Terraform syntax rules.

Looking ahead...

I'm thinking that `Callers` could also be checked for annotations, to preserve the existing behavior that allows this:

```tf
module "instance" {
  source = "./modules/instance"

  # tflint-ignore: aws_instance_invalid_type
  type = "t1.2xlarge"
}
```

I also think this needs to be supported if `--modules` does more than just argument validation:

```tf
# tflint-ignore: terraform_naming_convention
module "m" {
  source = "terraform-aws-modules/security-group/aws"

  # ...
}
```

This allows you to have style rules but optionally disable them (or TFLint) for entire third party modules.

And then we could revert to using the original source range from the child module. Perhaps `Callers` could be pretty-printed in the same way as the main source range, illustrating both the caller's input (`type = "t1.xlarge"`) and the raised expression (`instance_type = var.type`).

Sound reasonable @wata727?

**Comment by wata727** [comment]
Umm... I thought it was better to skip the inspection of the child modules in the `terraform_deprecated_interpolation` rule.

As you say, module inspection reports an issue due to the module arguments. This works as expected for rules such as `aws_instance_invalid_type`, but not for rules such as `terraform_deprecated_interpolation`, which has no cause in module arguments.

On the other hand, given the current behavior that running `tflint` on the top directory only checks the root module, there is a way to ignore the warning about deprecated interpolations in the child modules. To find deprecated interpolations inside child modules, you can run `tflint` for each module.

What do you think? @bendrucker

**Comment by bendrucker** [comment]
This would seem to apply to all `terraform_` rules, so we'd need to address this in a way that doesn't have every rule re-implement the same logic. Pretty sure this is coming up for `terraform_deprecated_interpolation` because it's commonly violated and enabled by default. It could work to only consider issues for rules where `Severity() == tflint.ERROR`, with a user configurable option to return lower levels.

**Comment by bendrucker** [comment]
Also having trouble seeing how 0.13 could have affected this. I haven't gone back and tried to reproduce it, but the mechanics of `EmitIssue()` around modules and variables haven't changed.
